xcodebuild/utils.go

```diff
@@ -0,0 +1,383 @@
+	}
+
+	switch params.TestRepetitionMode {
+	case untilFailure:
+		xcodebuildArgs = append(xcodebuildArgs, "-run-tests-until-failure")
+	case retryOnFailure:
+		xcodebuildArgs = append(xcodebuildArgs, "-retry-tests-on-failure")
+	}
+
+	if params.TestRepetitionMode != none {
+		xcodebuildArgs = append(xcodebuildArgs, "-test-iterations", strconv.Itoa(params.MaximumTestRepetitions))
+	}
+
+	if params.RelaunchTestsForEachRepetition {
+		xcodebuildArgs = append(xcodebuildArgs, "-test-repetition-relaunch-enabled", "YES")
+	}
+
+	if params.AdditionalOptions != "" {
+		options, err := shellquote.Split(params.AdditionalOptions)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse additional options (%s), error: %s", params.AdditionalOptions, err)
+		}
+		xcodebuildArgs = append(xcodebuildArgs, options...)
+	}
+
+	return xcodebuildArgs, nil
+}
+
+func (b *xcodebuild) createXCPrettyArgs(options string) ([]string, error) {
+	var args []string
+
+	if options != "" {
+		options, err := shellquote.Split(options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse additional options (%s), error: %s", options, err)
+		}
+		// get and delete the xcpretty output file, if exists
+		xcprettyOutputFilePath := ""
+		isNextOptOutputPth := false
+		for _, aOpt := range options {
+			if isNextOptOutputPth {
+				xcprettyOutputFilePath = aOpt
+				break
+			}
+			if aOpt == "--output" {
+				isNextOptOutputPth = true
+				continue
+			}
+		}
+		if xcprettyOutputFilePath != "" {
+			if isExist, err := b.pathChecker.IsPathExists(xcprettyOutputFilePath); err != nil {
+				b.logger.Errorf("Failed to check xcpretty output file status (path: %s), error: %s", xcprettyOutputFilePath, err)
+			} else if isExist {
+				b.logger.Warnf("=> Deleting existing xcpretty output: %s", xcprettyOutputFilePath)
+				if err := b.fileRemover.Remove(xcprettyOutputFilePath); err != nil {
+					b.logger.Errorf("Failed to delete xcpretty output file (path: %s), error: %s", xcprettyOutputFilePath, err)
+				}
+			}
+		}
+		//
+		args = append(args, options...)
+	}
+
+	return args, nil
+}
+
+func (b *xcodebuild) runTest(params TestRunParams) (string, int, error) {
+	xcodebuildArgs, err := createXcodebuildTestArgs(params.BuildTestParams, params.XcodeMajorVersion)
+	if err != nil {
+		return "", 1, err
+	}
+
+	b.logger.Infof("Running the tests...")
+
+	var rawOutput string
+	var exit int
+	var testErr error
+	if params.OutputTool == XcprettyTool {
+		xcprettyArgs, err := b.createXCPrettyArgs(params.XcprettyOptions)
+		if err != nil {
+			return "", 1, err
+		}
+
+		rawOutput, exit, testErr = b.runPrettyXcodebuildCmd(true, xcprettyArgs, xcodebuildArgs)
+	} else {
+		rawOutput, exit, testErr = b.runXcodebuildCmd(xcodebuildArgs...)
+	}
+
+	fmt.Println("exit: ", exit)
+
+	if testErr != nil {
+		return b.handleTestRunError(params, testRunResult{xcodebuildLog: rawOutput, exitCode: exit, err: testErr})
+	}
+
+	return rawOutput, exit, nil
+}
+
+type testRunResult struct {
```

Comment:
Can we move this to the top of the file? I almost missed it.

Reply:
Isn't it better to keep it close to its usage? If we move to the top, the first usage will come ~200 lines later.

Reply:
I'm okay with that too?